Quantum-Control-Applications/NV-Centers/Confocal-Setup/ramsey.py

```diff
@@ -0,0 +1,114 @@
+    n_st = declare_stream()  # stream to save iterations
+
+    play("laser_ON", "AOM")
+    wait(100)
+    with for_(n, 0, n < n_avg, n + 1):
+        with for_(t, t_min, t <= t_max, t + dt):
+            play("pi_half", "NV")  # Pi/2 pulse to qubit
+            wait(t, "NV")  # variable delay in spin Echo
+            play("pi_half", "NV")  # Pi/2 pulse to qubit
+
+            align()
+
+            play("laser_ON", "AOM")
+            measure("readout", "SPCM", None, time_tagging.analog(times1, meas_len, counts1))
+            save(counts1, counts_1_st)  # save counts
+            wait(100, "AOM")
+
+            align()
+
+            play("pi_half", "NV")  # Pi/2 pulse to qubit
+            wait(t, "NV")  # variable delay in spin Echo
+            frame_rotation_2pi(0.5, "NV")  # Turns next pulse to -x
+            play("pi_half", "NV")  # Pi/2 pulse to qubit
+            reset_frame("NV")
+
+            align()
+
+            play("laser_ON", "AOM")
+            measure("readout", "SPCM", None, time_tagging.analog(times2, meas_len, counts2))
+            save(counts2, counts_2_st)  # save counts
+            wait(100, "AOM")
+
+        save(n, n_st)  # save number of iteration inside for_loop
+
+    with stream_processing():
+        counts_1_st.buffer(len(t_vec)).average().save("counts1")
+        counts_2_st.buffer(len(t_vec)).average().save("counts2")
+        n_st.save("iteration")
+
+#####################################
+#  Open Communication with the QOP  #
+#####################################
+qmm = QuantumMachinesManager(host=qop_ip, port=qop_port)
+
+qm = qmm.open_qm(config)
+
+job = qm.execute(ramsey)  # execute QUA program
+
+res_handles = job.result_handles  # get access to handles
+counts1_handle = res_handles.get("counts1")
+counts2_handle = res_handles.get("counts2")
+iteration_handle = res_handles.get("iteration")
+counts1_handle.wait_for_values(1)
+counts2_handle.wait_for_values(1)
+iteration_handle.wait_for_values(1)
+
+
+def on_close(event):
+    event.canvas.stop_event_loop()
+    job.halt()
+
+
+f = plt.figure()
+f.canvas.mpl_connect("close_event", on_close)
+next_percent = 0.1  # First time print 10%
+print("Progress =", end=" ")
+
+b_cont = res_handles.is_processing()
+b_last = not b_cont
+
+while b_cont or b_last:
+    plt.cla()
+    counts1 = counts1_handle.fetch_all()
+    counts2 = counts2_handle.fetch_all()
+    iteration = iteration_handle.fetch_all() + 1
+    if iteration / n_avg > next_percent:
+        percent = 10 * round(iteration / n_avg * 10)  # Round to nearest 10%
+        print(f"{percent}%", end=" ")
+        next_percent = percent / 100 + 0.1  # Print every 10%
+
+    plt.plot(4 * t_vec, counts1 / 1000 / (meas_len * 1e-9), counts2 / 1000 / (meas_len * 1e-9))
+    plt.xlabel("Tau [ns]")
+    plt.ylabel("Intensity [kcps]")
+    plt.title("Ramsey")
+    plt.pause(0.1)
+
+    b_cont = res_handles.is_processing()
+    b_last = not (b_cont or b_last)
+
+print("")
```

Comment:
Just out of curiosity, why do you always measure along X and -X, is it standard for NV centers ?

Reply:
In room temperature NV measurements you get photons from both the |0> and |+-1> states. So comparing the two projections is a good indication for the state contrast compared to a Rabi measurement. In addition, this is a good way to differentiate between features (like an echo revival, or a decoherence peak),  and noise.

Reply:
I see, thanks !